### PR TITLE
Lesson Group: Group Lessons on Script Overview Page By Lesson Group

### PR DESCRIPTION
**Pull request by dmcavoy**
Makes it so if we have a lesson groups for lessons we use that to group the lessons instead of flex categories. If no lesson group is given and a flex category is we will continue to use the flex category.

I checked a 2019 script for csf, csd, csp, and a plc course. They all look right. Below are screenshots.

### CSP
Screen Shot 2020-04-22 at 4 36 45 PM

### CSD
Screen Shot 2020-04-22 at 4 36 38 PM

### CSF
Screen Shot 2020-04-22 at 4 36 31 PM

### Deeper Learning for Facilitators
Screen Shot 2020-04-22 at 4 36 22 PM

## Links

- [spec](https://docs.google.com/document/d/1Xt4dTTMGi3i4g5IhM8RpYXvyJr2KHsg9aoDw328LqMU/edit)
- [jira](https://codedotorg.atlassian.net/browse/PLAT-43)

## Testing story

Added tests in progressReduxTest for `groupedLessons` to check that it works in cases where lesson groups and flex categories are both specified or if only lesson group is specified. (Already had a test for just flex category)

# Reviewer Checklist:

- [ ] Tests provide adequate coverage
- [ ] Code is well-commented
- [ ] New features are translatable or updates will not break translations
- [ ] Relevant documentation has been added or updated
- [ ] User impact is well-understood and desirable
- [ ] Pull Request is labeled appropriately
- [ ] Follow-up work items (including potential tech debt) are tracked and linked

### CODE REVIEW

File: apps/src/code-studio/progressRedux.js [hunk 1]
```diff
@@ -636,48 +635,50 @@ export function statusForLevel(level, levelProgress) {
...
+  let byGroup = {};
 
   const allLevels = levelsByLesson(state);
 
-  state.stages.forEach((stage, index) => {
-    const category = stage.flex_category;
-    const lesson = lessonFromStageAtIndex(state, index);
-    let stageLevels = allLevels[index];
+  state.stages.forEach((lesson, index) => {
+    const group = lesson.lesson_group_name
```
Comment (dmcavoy) [inline apps/src/code-studio/progressRedux.js]: New lines 650 - 652 are really the key to this PR. This makes it so if we have a lesson group for a stage we group by that instead of the flex category. Otherwise we will continue to use the flex category. The rest of the changes here are just naming updates to move toward the world we want.

File: dashboard/app/models/plc/learning_module.rb [hunk 2]
```diff
@@ -46,8 +46,4 @@ class Plc::LearningModule < ActiveRecord::Base
   def required?
     module_type == REQUIRED_MODULE
   end
-
-  def name_with_required_tag
```
Comment (dmcavoy) [inline dashboard/app/models/plc/learning_module.rb]: Found this dead code when I was looking at how flex category impacts this model so I deleted it

File: apps/src/code-studio/progressRedux.js [hunk 3]
```diff
@@ -636,48 +635,50 @@ export function statusForLevel(level, levelProgress) {
...
+  let byGroup = {};
 
   const allLevels = levelsByLesson(state);
 
-  state.stages.forEach((stage, index) => {
-    const category = stage.flex_category;
-    const lesson = lessonFromStageAtIndex(state, index);
-    let stageLevels = allLevels[index];
+  state.stages.forEach((lesson, index) => {
+    const group = lesson.lesson_group_display_name
```
Comment (dmcavoy) [inline apps/src/code-studio/progressRedux.js]: This is the key change to this file. Everything else is just renaming. This allows us to live in a world where we can use flex category or lesson group until we have moved completely over to lesson group

Comment (davidsbailey) [reply apps/src/code-studio/progressRedux.js]: @dmcavoy does this mean that if we found that we broke something, we would only revert the change to this line to go back to using flex categories?

Comment (dmcavoy) [reply apps/src/code-studio/progressRedux.js]: Yes

Comment (davidsbailey) [reply apps/src/code-studio/progressRedux.js]: now that everything will already have a lesson group, does using the conditional buy us anything there? I would suggest, just making this always use the lesson group display name. or, if necessary, you could make it conditional upon an experiment first, so that we could verify in production that everything still looks ok using the experiment flag, and then "launch" the experiment as a separate step.

File: apps/test/unit/code-studio/progressReduxTest.js [hunk 4]
```diff
@@ -1076,16 +1077,112 @@ describe('progressReduxTest', () => {
...
+      assert.equal(groups[0].levels[0].length, 0);
+
+      groups = groupedLessons(state, true);
+      assert.equal(groups.length, 1);
+      assert.equal(groups[0].levels.length, 1);
+      assert.equal(groups[0].levels[0].length, 1);
+      assert.equal(groups[0].levels[0][0]['bonus'], true);
+    });
+
+    it('returns lesson group if a lesson has both lesson group and flex category', () => {
```
Comment (davidsbailey) [inline apps/test/unit/code-studio/progressReduxTest.js]: great tests! If we do end up having to undo this change for any reason, and we tried to do it by just changing the one line in progressRedux.js, then I would imagine some of these tests might break, since they presumably depend on the new behavior. if that's true, then I would imagine we'd want to just roll back the entire PR, rather than trying to fix a single line.

**Review by davidsbailey — COMMENTED** [review]
Hey @dmcavoy , the code in this PR looks great, however I'm balking a bit at the size of the PR vs. the urgency with which we'd want to be able to fix this if something went sideways. having stages appear out of order (or worse) would be a big disruption for classrooms.

Based on your updates to the .script files, I don't think the _risk_ per se is super high of something going wrong, it's mostly just the potential severity if something does go wrong times the chance that we'd run into problems rolling it back.

There are a few ways we could proceed, ranked from least to most cautious:
1. try to harder convince ourselves we have good UI test coverage + low risk + easy rollback path, and then ship this PR as-is
2. split this PR into smaller PRs, so that we have fewer lines of code in the PR that makes the change over to using lesson groups
3. keep this PR as-is, but use an experiment to control whether lesson groups or flex categories are used (as mentioned below)
4. try to do some kind of rigorous side-by-side comparison of what we're rendering (probably overkill)

I would lean toward (3), since it requires the least amount of surgery on this PR, and will make it easy to spot check a few script overview pages via side-by-side comparison. I am also open to 1 or 2. happy to discuss but wanted to get these thoughts down. sorry to be a drag... let's discuss tomorrow.
